test/OpenTelemetry.Tests/Logs/LogRecordTest.cs

```diff
@@ -0,0 +1,353 @@
+#else
+            this.loggerFactory = LoggerFactory.Create(builder =>
+#endif
+            {
+                builder.AddOpenTelemetry(options => options
+                    .AddProcessor(this.processor));
+                builder.AddFilter(typeof(LogRecordTest).FullName, LogLevel.Trace);
+            });
+
+#if NETCOREAPP2_1
+            this.serviceProvider = serviceCollection.BuildServiceProvider();
+            this.logger = this.serviceProvider.GetRequiredService<ILogger<LogRecordTest>>();
+#else
+            this.logger = this.loggerFactory.CreateLogger<LogRecordTest>();
+#endif
+        }
+
+        [Fact]
+        public void CheckCateogryNameForLog()
+        {
+            this.logger.LogInformation("Log");
+            var categoryName = this.exportedItems[0].CategoryName;
+
+            Assert.Equal(typeof(LogRecordTest).FullName, categoryName);
+        }
+
+        [Theory]
+        [InlineData(LogLevel.Trace)]
+        [InlineData(LogLevel.Debug)]
+        [InlineData(LogLevel.Information)]
+        [InlineData(LogLevel.Warning)]
+        [InlineData(LogLevel.Error)]
+        [InlineData(LogLevel.Critical)]
+        public void CheckLogLevel(LogLevel logLevel)
+        {
+            var message = $"Log {logLevel}";
+            this.logger.Log(logLevel, message);
+
+            var logLevelRecorded = this.exportedItems[0].LogLevel;
+            Assert.Equal(logLevel, logLevelRecorded);
+        }
+
+        [Fact]
+        public void CheckStateForUnstructuredLog()
+        {
+            var message = "Hello, World!";
+            this.logger.LogInformation(message);
+            var state = this.exportedItems[0].State;
+            var itemCount = state.GetType().GetProperty("Count").GetValue(state);
+
+            // state only has {OriginalFormat}
+            Assert.Equal(1, itemCount);
+
+            Assert.Equal(message.ToString(), state.ToString());
+        }
+
+        [Fact]
+        public void CheckStateForUnstructuredLogWithStringInterpolation()
+        {
+            var message = $"Hello from potato {0.99}.";
+            this.logger.LogInformation(message);
+            var state = this.exportedItems[0].State;
+            var itemCount = state.GetType().GetProperty("Count").GetValue(state);
+
+            // state only has {OriginalFormat}
+            Assert.Equal(1, itemCount);
+
+            Assert.Equal(message.ToString(), state.ToString());
+        }
+
+        [Fact]
+        public void CheckStateForStructuredLogWithTemplate()
+        {
+            this.logger.LogInformation("Hello from {name} {price}.", "tomato", 2.99);
+            var state = this.exportedItems[0].State;
+            var itemCount = state.GetType().GetProperty("Count").GetValue(state);
+
+            // state has name, price and {OriginalFormat}
+            Assert.Equal(3, itemCount);
+
+            // Get value for {name}
+            var firstArgument = state.GetType().GetProperty("Item").GetValue(state, new object[] { 0 });
```

Comment:
we can avoid reflection here. Check the following:
```
var stateDictionary = (IReadOnlyCollection<KeyValuePair<string, object>>)logRecord.State;
foreach (var entry in stateDictionary )
{
}
```

Reply:
Fixed this,